### PR TITLE
Afh table

### PR DESCRIPTION
**Pull request by mythicat554**

### CODE REVIEW

File: constellations/orion/src/domains/ask_for_help.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	UpdatedAt  time.Time `json:"-" db:"updated_at"`
+	DeletedAt  NullTime  `json:"-" db:"deleted_at"`
+	Title      string    `json:"title"`
+	Date       string    `json:"date"`
+	TimeString string    `json:"timeString"`
+	Subject    string    `json:"subject"`
+	LocationId string    `json:"locationId" db:"location_id"`
+}
+
+func (askForHelp *AskForHelp) Validate() error {
```
Comment (tonywu315) [inline constellations/orion/src/domains/ask_for_help.go:21]: You should validate something at least. For example, `Title` should include at least 1 letter.

File: constellations/orion/src/domains/ask_for_help_valid_test.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package domains_test
+
+import (
+	"testing"
+
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/domains"
+)
+
+func TestValidAFHTitle(t *testing.T) {
+	askForHelp := domains.AskForHelp{}
```
Comment (tonywu315) [inline constellations/orion/src/domains/ask_for_help_valid_test.go:10]: Add actual values

File: constellations/orion/src/repos/migrations/000019_create_table_ask_for_help.up.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE ask_for_help
+(
+    id          int unsigned NOT NULL AUTO_INCREMENT,
+    created_at  datetime     NOT NULL,
+    updated_at  datetime     NOT NULL,
+    deleted_at  datetime,
+    title       text         NOT NULL,
```
Comment (tonywu315) [inline constellations/orion/src/repos/migrations/000019_create_table_ask_for_help.up.sql:7]: It would probably be better if this is `varchar(256)` instead of `text` to save space.

File: constellations/orion/src/repos/migrations/000019_create_table_ask_for_help.up.sql [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    date        date         NOT NULL,
+    timeString  varchar(64)  NOT NULL,
+    subject     text         NOT NULL,
```
Comment (tonywu315) [inline constellations/orion/src/repos/migrations/000019_create_table_ask_for_help.up.sql:10]: This is even shorter so `varchar(128)` is enough.

File: constellations/orion/src/repos/ask_for_help_repo.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
+package repos
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/ahsu1230/mathnavigatorSite/constellations/orion/src/repos/utils"
+
```
Comment (tonywu315) [inline constellations/orion/src/repos/ask_for_help_repo.go:8]: Unnecessary whitespace

File: constellations/orion/src/repos/ask_for_help_repo_test.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	db, mock, repo := initAFHTest(t)
+	defer db.Close()
+
+	// Mock DB statements and execute
+	now := time.Now().UTC()
+	var date1 = now.Add(time.Hour * 24 * 30)
+	result := sqlmock.NewResult(1, 1)
+	mock.ExpectPrepare("^INSERT INTO ask_for_help").
+		ExpectExec().
+		WithArgs(sqlmock.AnyArg(), sqlmock.AnyArg(), "AP Calculus Help", date1, "2:00-4:00 PM", "AP Calculus", "wchs").
```
Comment (tonywu315) [inline constellations/orion/src/repos/ask_for_help_repo_test.go:83]: This line is really long and should be split into several lines.

File: constellations/orion/src/repos/ask_for_help_repo_test.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	db, mock, repo := initAFHTest(t)
+	defer db.Close()
+
+	// Mock DB statements and execute
+	now := time.Now().UTC()
+	var date1 = now.Add(time.Hour * 24 * 30)
+	result := sqlmock.NewResult(1, 1)
+	mock.ExpectPrepare("^UPDATE ask_for_help SET (.*) WHERE id=?").
+		ExpectExec().
+		WithArgs(sqlmock.AnyArg(), 2, "AP Stat Help", date1, "2:00-4:00PM", "AP Stat", "room12", 1).
```
Comment (tonywu315) [inline constellations/orion/src/repos/ask_for_help_repo_test.go:113]: Same

File: constellations/orion/src/repos/ask_for_help_repo_test.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+	// Validate results
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("Unfulfilled expectations: %s", err)
+	}
+}
+
+// Helper Methods
+func getAFHRows() *sqlmock.Rows {
+	return sqlmock.NewRows([]string{"Id", "CreatedAt", "UpdatedAt", "DeletedAt", "Title", "Date", "TimeString", "Subject", "LocationId"}).
```
Comment (tonywu315) [inline constellations/orion/src/repos/ask_for_help_repo_test.go:157]: Same

File: constellations/orion/src/tests_integration/ask_for_help_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+	utils.ResetTable(t, domains.TABLE_ASKFORHELP)
+	utils.ResetTable(t, domains.TABLE_LOCATIONS)
+}
+
+// Test: Create 1 Ask For Help, Update, Get By ID
+func Test_UpdateAFH(t *testing.T) {
+	createLocations(t)
+
+	//Create 1 AFH
```
Comment (tonywu315) [inline constellations/orion/src/tests_integration/ask_for_help_integration_test.go:70]: Add a space after //

File: constellations/orion/go.sum [hunk 10]
```diff
@@ -4,6 +4,8 @@ github.com/Microsoft/go-winio v0.4.14 h1:+hMXMk01us9KgxGb7ftKQt2Xpf5hH/yky+TDA+q
 github.com/Microsoft/go-winio v0.4.14/go.mod h1:qXqCSQ3Xa7+6tgxaGTIe4Kpcdsi+P8jBhyzoq1bpyYA=
```
[the remark this answers is not in the document]
Comment (mythicat554) [reply constellations/orion/go.sum:4]: Yea, I kind of just rolled with it. I'm not sure why it appeared either?

File: constellations/orion/src/controllers/ask_for_help_controller_test.go [hunk 11]
```diff
@@ -24,8 +22,21 @@ var date2 = now.Add(time.Hour * 24 * 31)
...
-			testUtils.CreateMockAFH(2, "AP Statistics Help", date2, "3:00-5:00PM", "AP Statistics", "room12"),
+			testUtils.CreateMockAFH(
+				1,
+				"AP Calculus Help",
+				date1,
+				"2:00-4:00PM",
+				"AP Calculus",
+				"wchs",
+			),
+			testUtils.CreateMockAFH(2,
```
Comment (tonywu315) [inline constellations/orion/src/controllers/ask_for_help_controller_test.go:33]: The 2 is on the same line

Comment (mythicat554) [reply constellations/orion/src/controllers/ask_for_help_controller_test.go]: Oops, my bad. Thanks for reviewing!

File: constellations/orion/go.sum (hunk 10 above)
[the remark this answers is not in the document]
Comment (ahsu1230) [reply constellations/orion/go.sum:4]: go.sum is fine to checkin. https://www.reddit.com/r/golang/comments/97dxj1/should_i_add_the_gosum_file_to_the_repo_of_a/

It's used to verify packages are essentially correctly installed

File: constellations/orion/src/domains/ask_for_help.go [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+	Id         uint      `json:"id"`
+	CreatedAt  time.Time `json:"-" db:"created_at"`
+	UpdatedAt  time.Time `json:"-" db:"updated_at"`
+	DeletedAt  NullTime  `json:"-" db:"deleted_at"`
+	Title      string    `json:"title"`
+	Date       time.Time `json:"date"`
+	TimeString string    `json:"timeString" db:"time_string"`
+	Subject    string    `json:"subject"`
+	LocationId string    `json:"locationId" db:"location_id"`
+}
```
[the remark this answers is not in the document]
Comment (ahsu1230) [reply constellations/orion/src/domains/ask_for_help.go:22]: Sorry, the type of Notes should be NullString! You're migration file is good though.

[the remark this answers is not in the document]
Comment (ahsu1230) [reply constellations/orion/src/domains/ask_for_help.go:22]: Look for how other places in the codebase use `domains.NullString`

[the remark this answers is not in the document]
Comment (mythicat554) [reply constellations/orion/src/domains/ask_for_help.go:22]: Oops, I didn't realize you wanted NullString! (ig I'm just blind)
I just fixed it.
